[BUG] Query failure with java.lang.AssertionError when using partitioned Iceberg tables

**Describe the bug**
When running the NDS 2.0 Power Run locally using partitioned Iceberg tables with Parquet data, the following exception is encountered when query 7 is run:

```
Caused by: java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:208)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$doExecuteBroadcast$1(AdaptiveSparkPlanExec.scala:359)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.withFinalPlanUpdate(AdaptiveSparkPlanExec.scala:366)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.doExecuteBroadcast(AdaptiveSparkPlanExec.scala:358)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeBroadcast$1(SparkPlan.scala:197)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:222)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:219)
	at org.apache.spark.sql.execution.SparkPlan.executeBroadcast(SparkPlan.scala:193)
	at org.apache.spark.sql.execution.SubqueryBroadcastExec.$anonfun$relationFuture$2(SubqueryBroadcastExec.scala:81)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withExecutionId$1(SQLExecution.scala:139)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
	at org.apache.spark.sql.execution.SQLExecution$.withExecutionId(SQLExecution.scala:137)
	at org.apache.spark.sql.execution.SubqueryBroadcastExec.$anonfun$relationFuture$1(SubqueryBroadcastExec.scala:78)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
```

**Steps/Code to reproduce bug**
Create a partitioned table using Iceberg parquet data and run NDS query 7 against it using the Spark RAPIDS plugin (latest 22.06 version).

**Expected behavior**
Expectation is that query will succeed and not fail with an exception.

**Environment details (please complete the following information)**
Environment is local Linux desktop with RTX A5000 GPU.  Spark version is 3.2.1 with standard benchmark configuration.

**Additional context**
N/A

--- a/integration_tests/src/main/python/iceberg_test.py
+++ b/integration_tests/src/main/python/iceberg_test.py
@@ -12,9 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import pytest
+
 from asserts import assert_gpu_and_cpu_are_equal_collect
-from marks import allow_non_gpu, iceberg
-from spark_session import with_cpu_session
+from data_gen import *
+from marks import allow_non_gpu, iceberg, ignore_order
+from spark_session import is_before_spark_320, is_databricks_runtime, with_cpu_session
 
 @allow_non_gpu('BatchScanExec')
 @iceberg
@@ -26,3 +29,21 @@ def test_iceberg_fallback_not_unsafe_row(spark_tmp_table_factory):
     with_cpu_session(setup_iceberg_table)
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark : spark.sql("SELECT COUNT(DISTINCT id) from {}".format(table)))
+
+@allow_non_gpu('BatchScanExec')
+@iceberg
+@ignore_order(local=True)
+@pytest.mark.skipif(is_before_spark_320() or is_databricks_runtime(),
+                    reason="AQE+DPP not supported until Spark 3.2.0+ and AQE+DPP not supported on Databricks")
+def test_iceberg_aqe_dpp(spark_tmp_table_factory):
+    table = spark_tmp_table_factory.get()
+    def setup_iceberg_table(spark):
+        df = two_col_df(spark, int_gen, int_gen)
+        df.createOrReplaceTempView("df")
+        spark.sql("CREATE TABLE {} (a INT, b INT) USING ICEBERG PARTITIONED BY (a)".format(table))
+        spark.sql("INSERT INTO {} SELECT * FROM df".format(table))
+    with_cpu_session(setup_iceberg_table)
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark : spark.sql("SELECT * from {} as X JOIN {} as Y ON X.a = Y.a WHERE Y.a > 0".format(table, table)),
+        conf={"spark.sql.adaptive.enabled": "true",
+              "spark.sql.optimizer.dynamicPartitionPruning.enabled": "true"})

--- a/sql-plugin/src/main/320+/scala/com/nvidia/spark/rapids/shims/Spark320PlusShims.scala
+++ b/sql-plugin/src/main/320+/scala/com/nvidia/spark/rapids/shims/Spark320PlusShims.scala
@@ -492,6 +492,33 @@ trait Spark320PlusShims extends SparkShims with RebaseShims with Logging {
       parent: Option[RapidsMeta[_, _, _]],
       rule: DataFromReplacementRule)
       extends SparkPlanMeta[BatchScanExec](p, conf, parent, rule) {
+    // Replaces SubqueryBroadcastExec inside dynamic pruning filters with GPU counterpart
+    // if possible. Instead regarding filters as childExprs of current Meta, we create
+    // a new meta for SubqueryBroadcastExec. The reason is that the GPU replacement of
+    // BatchScanExec is independent from the replacement of the runtime filters. It is
+    // possible that the BatchScanExec is on the CPU, while the dynamic runtime filters
+    // are on the GPU. And vice versa.
+    private lazy val runtimeFilters = {
+      val convertBroadcast = (bc: SubqueryBroadcastExec) => {
+        val meta = GpuOverrides.wrapAndTagPlan(bc, conf)
+        meta.tagForExplain()
+        meta.convertIfNeeded().asInstanceOf[BaseSubqueryExec]
+      }
+      wrapped.runtimeFilters.map { filter =>
+        filter.transformDown {
+          case dpe @ DynamicPruningExpression(inSub: InSubqueryExec) =>
+            inSub.plan match {
+              case bc: SubqueryBroadcastExec =>
+                dpe.copy(inSub.copy(plan = convertBroadcast(bc)))
+              case reuse @ ReusedSubqueryExec(bc: SubqueryBroadcastExec) =>
+                dpe.copy(inSub.copy(plan = reuse.copy(convertBroadcast(bc))))
+              case _ =>
+                dpe
+            }
+        }
+      }
+    }
+
     override val childScans: scala.Seq[ScanMeta[_]] =
       Seq(GpuOverrides.wrapScan(p.scan, conf, Some(this)))
 
@@ -501,7 +528,11 @@ trait Spark320PlusShims extends SparkShims with RebaseShims with Logging {
       }
     }
 
+    override def convertToCpu(): SparkPlan = {
+      wrapped.copy(runtimeFilters = runtimeFilters)
+    }
+
     override def convertToGpu(): GpuExec =
-      GpuBatchScanExec(p.output, childScans.head.convertToGpu(), p.runtimeFilters)
+      GpuBatchScanExec(p.output, childScans.head.convertToGpu(), runtimeFilters)
   }
 }

